frb/dm/mcmc.py

```diff
@@ -0,0 +1,320 @@
+""" Methods for MCMC analysis of the Macquart relation """
+from numba.core.typeinfer import DelItemConstraint
+import numpy as np
+from numba import njit, prange
+
+from scipy.stats import lognorm
+from scipy.interpolate import InterpolatedUnivariateSpline as IUS
+
+import warnings
+
+try:
+    import pymc3 as pm
+except ImportError:
+    warnings.warn("You need to install pymc3 to run mcmc codes")
+else:
+    import theano.tensor as tt
+    from theano.compile.ops import as_op
+
+from astropy.cosmology import Planck15
+
+from frb.dm import cosmic, igm
+
+from IPython import embed
+
+DM_values = np.linspace(1., 7000., 7000)
+Delta_values = np.linspace(1./400., 20., 20000)# / 400.  # Fiducial
+DM_MWhalo = 50.  # Fixed value for Macquart+20
+
+# Init as blank
+frbs = []
+frb_DMs = None #np.array([frb.DM.value-frb.DMISM.value for frb in frbs])
+frb_zs = None #np.array([frb.z for frb in frbs])
+DM_FRBp_grid = None
+DMhost_grid = None #np.outer(DM_values, (1+z_FRB))  # Host rest-frame DMs
+DMvalues_grid = None # np.outer(DM_values, np.ones(DM_FRBp.size))
+Deltavalues_grid = None  #np.outer(Delta_values, np.ones_like(C0)), 
+
+#
+
+# DM Cosmic
+DM_cosmic, zeval = igm.average_DM(1., cosmo=Planck15, cumul=True)
+spl_DMc = IUS(zeval, DM_cosmic.value)
+Planck15_ObH0 = Planck15.Ob0 * Planck15.H0.value
+Planck15_Obh70 = Planck15.Ob0 * (Planck15.H0.value/70.)
+
+# Load splines
+spl_sigma = cosmic.grab_sigma_spline()
+f_C0_3 = cosmic.grab_C0_spline()
+
+
+# Parameters
+def grab_parmdict(tight_ObH=False):
+    """ Generate the parameter dict for the MCMC run
+
+    Args:
+        tight_ObH (bool, optional): [description]. Defaults to False.
+
+    Raises:
+        IOError: [description]
+
+    Returns:
+        dict: [description]
+    """
+    parm_dict = {}
+    parm_dict['F'] = dict(dist='Uniform', lower=0.011, upper=0.5, latex='F', unit='None')
+    parm_dict['mu'] = dict(dist='Uniform', lower=20., upper=200, latex='\exp(\mu)', unit='pc cm$^{-3}$')
+    parm_dict['lognorm_s'] = dict(dist='Uniform', lower=0.2, upper=2, latex='\sigma_{\\rm host}', unit='pc $^{-3}$')
+    if tight_ObH:
+        # NEED TO FIX THIS for h70!
+        raise IOError
+        parm_dict['ObH0'] = dict(dist='Normal', mu=Planck15_ObH0,
+                                 sigma=Planck15_ObH0*0.03)
+    else:
+        parm_dict['Obh70'] = dict(dist='Uniform', lower=0.015, upper=0.095)
+    parm_dict['Obh70']['latex'] = '\\Omega_b h_{70}'
+    parm_dict['Obh70']['unit'] = 'None' #$\\rm km \, s^{-1} Mpc^{-1}$'
+    # Other info
+    parm_dict['Other'] = dict(DM_MWhalo=50., zscale_host=True, floor=0.)
+    # Return    
+    return parm_dict
+
+def one_prob(Obh70, F, DM_FRBp, z_FRB, mu=150., lognorm_s=1.,
+             lognorm_floor=0., orig=False, beta=4.):
+    """
+
+    Args:
```

Comment:
Args and return values require descriptions and data types.

Reply:
added